### ISSUE TITLE
Implement functionality to obfuscate 'entire project' in production export, along with its compatible filesystem

### ISSUE THREAD

**Issue by SysError99** [godotengine/godot-proposals#6395]
### Describe the project you are working on

Many of instant games made with Godot 3.x

### Describe the problem or limitation you are having in your project

Regarding #4220 and #1407, as they described as why we need GDScript obfuscation in production exports. However, as many had already mentioned that obfuscating scripts alone isn't really 'beneficial' since it still can be de-obfuscated and re-labelled quite easily considering how Godot works with GDScript (and also C#). But as I looked through some of game engines and and how their building process worked, I finally realised that we are focusing on the wrong path.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

To describe how some build processes of how build processes of other game engines. They usually present their own method to handle internal files that will ensure that every attempt of file accesses are compatible for project obfuscation. Then, they will try to obfuscate files and directories in the bundle to make sure that it cannot be reconstructed easily. This is very common practice in many of other game engines, such as GameMaker, GDevelop, Cocos, etc.

Which comes into my conclusion that we could have this functionality built into the game engine (at least, as an option and disable string-path-based file access entirely, since it's not compatible with obfuscation), and provide more reliable way to protect production exports as it cannot be altered easily.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Provide an option to make a project 'obfuscation-compatible' in project creation menu and project settings (forcing it to obfuscate all projects isn't suitable for everyone), along with new methods to access filesystem that is compatible with obfuscation, and prohibit common filesystem access entirely since it's not compatible anymore. Maybe we could have it in another internal file protocol, such as `ores://`, `bundle://`, etc.

However, this could means that certain implementations of resource transferring (such as, JSON serialisation/deserialisation) from external sources (such as REST APIs)  broken, which we could also provide internal mapping functions to address the issue, and also discourage the use of external formats to reduce complications in data alternation.

The strategy on how it obfuscates the project is rather simple, assuming that Godot already provide safe methods to access internal filesystems.
1. Note every single file path, obfuscate, and move them into root directory.
For example, we have this file structure.
```md
- images
    - player_sprite.png
    - enemy_sprite.png
    - tileset.png
- resources
    - tileset.tres
    - theme.tres
- objects
    - player.tscn
    - player.gd
    - enemy.tscn
    - enemy.gd
- project.godot
- main_menu.tscn
- main_menu.gd
- level_0.tscn
- level_0.gd
- level_1.tscn
- level_1.gd
- victory_screen.tscn
- victory_screen.gd
```
Can be obfuscated to:
```md
- a.png
- b.png
- c.png
- a.tres
- b.tres
- a.tscn
- b.tscn
- c.tscn
- d.tscn
- e.tscn
- f.tscn
- a.gd
- b.gd
- c.gd
- d.gd
- e.gd
- f.gd
- project.godot // This is entry point, cannot obfuscate its name.
```
2. Note all variables, internal labels, user-defined indexes, and obfuscate them, excluding certain labels such as translation (however translation keys can still be obfuscated)
3. (optional) move all strings from script and and store them in single script desiged for storing all script strings.
4. Export the obfuscated project as usual.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

I don't believe we can achieve obfuscation with few of extensions since it involves overhauling build process.

### Is there a reason why this should be core and not an add-on in the asset library?

I could make my own tool that will automate obfuscation, but I still need to be very cautious on file management since there is no indications to warn users whatsoever. Plus, there's also no reason of why Godot user projects shouldn't be protected by default.

**Comment by YuriSizov** [comment]
Script obfuscation has a specific goal to hide the logic behind the game for one reason or another. Minimizing names of project files doesn't help with that, and by itself doesn't seem to be terribly useful. And it will break your code, if it relies on path concatenation. And it will require the engine to somehow modify your scripts to replace all the paths, even if they are compile-time resolvable. All in all, this would be a lot of work for very little gain.

**Comment by SysError99** [comment]
The explanation could be unclear but things like just-minimising names aren't in part of this proposal. It can be anything obscure and if-not-so difficult to reconstruct project files, relabel, and make changes on it. Considering in theory it's little gain from big effort but we have already seen in other game engines that it's quite effective to stop casual hacks. It doesn't protect repacking, to be fair (even best obfuscation won't help), but it makes the alternation of codebase or even project itself much harder (the easiest they can do is making changes of resource files and making guesses on how the game works)

This proposal can be a little bit extended by randomising labels in every build, which discourages the effort to reconstruct updated version in every release.

> Minimizing names of project files doesn't help with that, and by itself doesn't seem to be terribly useful. And it will break your code, if it relies on path concatenation. And it will require the engine to somehow modify your scripts to replace all the paths, even if they are compile-time resolvable.

This is also part of this proposal to address the issue as we should also have another form of filesystem that will be used to access project resources, and discourage the use of common file path manipulation (which, by itself, also is quite risky and prone of error by itself). For example, GameMaker removes the aspect of file system entirely and simply use resource referencing, which helps solving the problem in this regard.

**Comment by YuriSizov** [comment]
You can already use UID paths in your code, but that's not great in terms of usability and if you work in teams. And reliance on the internal filesystem is a powerful tool for data-driven games and game systems.

Outside of code you already have UIDs for references and binary format for scenes and other resources for extra obscurity.

> The explanation could be unclear but things like just-minimising names aren't in part of this proposal.

Then what is? What do you imply by obfuscation other than minified file identifiers that you demonstrate? What do you call internal labels? How do you "move all the strings from scripts"? How does that help with obscurity?

**Comment by SysError99** [comment]
The step 3 on the obfuscation can be my mistake since I could be mistaken with `Macro Tool` in GameMaker. We could forget this part temporarily until I can come up with more solid definition.

I should also do a retake on the explanation. On the obfuscation part, what I meant is that we can re-label all of user-defined labels (such as, variables, functions, custom signals, classes, etc) in script files to random strings that aren't meaningful. We can also convert all file paths in supposedly 'restricted' filesystem (the same way as how ResourceUID works), and move them into project root directory so it's harder to identify files. Also, we need to reflect all randomised labels to all Godot files (TSCN, TRES) so it's still working with randomised labels in script side. I think these are already sufficient to give some obscurity to Godot project, in a sacrifice of very primitive resource management (as it no longer possible to use `res://` protocol).

**Comment by SysError99** [comment]
After few researches, I decided to drop this proposal since it's not really possible to obfuscate entire project without having significant drawbacks, and potentially will need entire project rework.

I decided to make a dedicated tool that will be designed to obfuscate the project, as it still has some of possibilities to get it working. However, there will be significant limitations that users will need to follow.

Closing this issue as not solved.
